enterprise/server/backends/prom/prom.go

```diff
@@ -0,0 +1,143 @@
+package prom
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/buildbuddy-io/buildbuddy/server/environment"
+	"github.com/buildbuddy-io/buildbuddy/server/metrics"
+	"github.com/buildbuddy-io/buildbuddy/server/util/log"
+	"github.com/buildbuddy-io/buildbuddy/server/util/status"
+	"github.com/prometheus/client_golang/api"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+
+	promapi "github.com/prometheus/client_golang/api/prometheus/v1"
+)
+
+var (
+	address = flag.String("prometheus.address", "", "the address of the promethus HTTP API")
+)
+
+type promQuerier struct {
+	api promapi.API
+}
+
+func newQuerier() (*promQuerier, error) {
+	if len(*address) == 0 {
+		return nil, nil
+	}
+	c, err := api.NewClient(api.Config{
+		Address: *address,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &promQuerier{
+		api: promapi.NewAPI(c),
+	}, nil
+}
+
+type bbMetricsCollector struct {
+	env             environment.Env
+	groupID         string
+	InvocationCount *prometheus.Desc
+}
+
+type metricsGroupRegistryManager struct {
+	env               environment.Env
+	registryByGroupID sync.Map // groupID -> registry
+}
+
+func newMetricsGroupRegistry(env environment.Env) *metricsGroupRegistryManager {
+	return &metricsGroupRegistryManager{
+		env:               env,
+		registryByGroupID: sync.Map{},
+	}
+}
+
+func Register(env environment.Env) error {
+	q, err := newQuerier()
+	if err != nil {
+		return status.InternalErrorf("failed to configure prom querier: %s", err)
+	}
+	env.SetPromQuerier(q)
+
+	m := newMetricsGroupRegistry(env)
+	env.SetMetricsGroupRegistries(m)
+	return nil
+}
+
+func (m *metricsGroupRegistryManager) GetOrCreateRegistry(groupID string) (*prometheus.Registry, error) {
```

Comment:
I don't know what the overhead is for a registry -- but if Registry is going to call collector.Collect any time metrics are scraped then is there any value in holding on to the Registry object between requests?

(especially given that subsequent requests are very likely to hit another server)

Could you instead dynamically create a Registry when the scape endpoint is hit, calling your collector (which hopefully hits cached data some of the time)

Reply:
Done. I don't think NewRegistry() itself has any overhead. However reg.Register(collector) has some overhead as it is trying to validate the collector.